mongo/description/server_selector.go

```diff
@@ -115,37 +166,66 @@ func (ls *latencySelector) SelectServer(t Topology, candidates []Server) ([]Serv
 	}
 }
 
-// WriteSelector selects all the writable servers.
-func WriteSelector() ServerSelector {
-	return ServerSelectorFunc(func(t Topology, candidates []Server) ([]Server, error) {
-		switch t.Kind {
-		case Single, LoadBalanced:
-			return candidates, nil
-		default:
-			result := []Server{}
-			for _, candidate := range candidates {
-				switch candidate.Kind {
-				case Mongos, RSPrimary, Standalone:
-					result = append(result, candidate)
-				}
+type writeServerSelector struct {
+	fn ServerSelectorFunc
+}
+
+func (writeServerSelector) info() serverSelectorInfo {
+	return serverSelectorInfo{Type: "writeSelector", CustomSelector: true}
+}
+
+func (selector writeServerSelector) String() string {
+	return selector.info().String()
+}
+
+func (selector writeServerSelector) SelectServer(t Topology, s []Server) ([]Server, error) {
+	return selector.fn(t, s)
+}
+
+func writeSelectorFunc(t Topology, candidates []Server) ([]Server, error) {
+	switch t.Kind {
+	case Single, LoadBalanced:
+		return candidates, nil
+	default:
+		result := []Server{}
+		for _, candidate := range candidates {
+			switch candidate.Kind {
+			case Mongos, RSPrimary, Standalone:
+				result = append(result, candidate)
 			}
-			return result, nil
 		}
-	})
+		return result, nil
+	}
 }
 
-// ReadPrefSelector selects servers based on the provided read preference.
-func ReadPrefSelector(rp *readpref.ReadPref) ServerSelector {
-	return readPrefSelector(rp, false)
+// WriteSelector selects all the writable servers.
+func WriteSelector() ServerSelector {
+	return writeServerSelector{
+		fn: writeSelectorFunc,
+	}
 }
 
-// OutputAggregateSelector selects servers based on the provided read preference given that the underlying operation is
-// aggregate with an output stage.
-func OutputAggregateSelector(rp *readpref.ReadPref) ServerSelector {
-	return readPrefSelector(rp, true)
+type readPrefServerSelector struct {
+	stringer fmt.Stringer
+	fn       ServerSelectorFunc
```

Comment:
The additional indirection added by using interface and func fields is unnecessary and can be replaced by simpler value fields and methods.

E.g.
```go
func ReadPrefSelector(rp *readpref.ReadPref) ServerSelector {
	return readPrefServerSelector{
		rp:                rp,
		isOutputAggregate: false,
	}
}

type readPrefServerSelector struct {
	rp                *readpref.ReadPref
	isOutputAggregate bool
}

func (selector readPrefServerSelector) info() serverSelectorInfo {
	return serverSelectorInfo{
		Type: "readPrefSelector",
		Data: selector.rp.String(),
	}
}

func (selector readPrefServerSelector) String() string {
	return selector.info().String()
}

func (selector readPrefServerSelector) SelectServer(...) ([]Server, error) {
	// ...
}
```

The same simplification can be made to all of the non-composite server selectors.